### ISSUE TITLE
OOM when importing trace from http://news.tfw2005.com/ with rendering categories

### ISSUE THREAD

**Issue by natduca** [catapult-project/catapult#2051]
From sunnyps@chromium.org on https://bugs.chromium.org/p/chromium/issues/detail?id=587598:

> I tried recording a trace of http://news.tfw2005.com/ using the "Rendering" categories. The tracing tool crashes when I finish the trace even though the buffer wasn't filled. Crash says that the tracing tab ran out of memory.

**Comment by natduca** [comment]
@petrcermak hm

**Comment by skyostil** [comment]
FWIW I seem to be hitting these kinds of OOM issues more frequently. I'll report back if I find a good repro.

**Comment by petrcermak** [comment]
We keep getting these reports quite often (e.g. https://bugs.chromium.org/p/chromium/issues/detail?id=587010), but triaging them is quite difficult without traces. It would be great if there was a way to __save a recorded trace without importing it__ so that the reporter could reproduce the issue (without crashing) and attach a trace *(feature proposal?)*.

**Comment by petrcermak** [comment]
This doesn't belong to the _Memory-Infra_ category (which refers to the "infrastructure" rather than "crashes"). I created a new label for these issues.

**Comment by petrcermak** [comment]
Reproduction details (from https://bugs.chromium.org/p/chromium/issues/detail?id=587598):

> Chrome ToT on Goobuntu
> 
> flags: --enable-smooth-scrolling --top-chrome-md=material
> gn args: is_component_build = true is_debug = false use_goma = true

**Comment by petrcermak** [comment]
### TL;DR ###

This is a __v8 and/or jszip issue__ and there's a __[simple fix](https://codereview.chromium.org/1739663003)__ for Catapult :tada:

### (Very Long) Explanation

I was able to consistently reproduce the issue with traces having buffer usage over ~40%. Using DevTools, I found that the OOM occurs in a call to <tt>JSZip.utils.transformTo</tt> [here](https://github.com/catapult-project/catapult/blob/c61947aa20dbfeda7920d531575fd1496dc5c8a9/tracing/tracing/extras/importer/gzip_importer.html#L116) in <tt>GzipImporter</tt>. The method converts the raw gzip inflated data (<tt>Uint8Array</tt>) to raw JSON (<tt>string</tt>).

```
GzipImporter.inflateGzipData_ = function(data) {
  // Gzip header handling.
  [...]

  // Inflate the data using jszip.
  // Raw deflated data -> raw inflated data (Uint8Array).
  var inflated_data = JSZip.compressions['DEFLATE'].uncompress(data.subarray(position));

  // Raw inflated data (Uint8Array) -> JSON string.
  // OOM OCCURS INSIDE THIS CALL.
  return JSZip.utils.transformTo('string', inflated_data);
},
```

This was very _surprising_ because the total size of the heap between inflation and conversion was _only_ 62.4 MiB (42.7 MiB was occupied by the <tt>Uint8Array</tt> containing the raw inflated data), which was _nowhere near_ the memory usage reported at the OOM (1425 MiB).

I re-ran the same experiment with a smaller trace which did _not_ OOM (~15% buffer usage) and found that the memory usage less than doubled after the <tt>JSZip.utils.transformTo</tt> call (32.5 MiB → 50.7 MiB).

These results indicated that __<tt>JSZip.utils.transformTo</tt> used lots of memory for intermediate operations__, so I started investigating memory usage inside the third-party method (using another small trace of similar size) and found that the total size of the heap right after the construction of the JSON string [here](https://github.com/Stuk/jszip/blob/3a39a715a5c522686aad95026134be3d5731d06e/lib/utils.js#L124) was _staggering_ 436 MiB (more than 13× and 9× increase compared to before (31.9 MiB) and after (44.5 MiB) the <tt>JSZip.utils.transformTo</tt> call, respectively)!

The jszip conversion method proceeds by [_(1)_ transforming 64 KiB chunks](https://github.com/Stuk/jszip/blob/3a39a715a5c522686aad95026134be3d5731d06e/lib/utils.js#L116) (65,536 characters) of the <tt>Uint8Array</tt> into strings using <tt>String.fromCharCode</tt> and then [_(2)_ joining all of them them](https://github.com/Stuk/jszip/blob/3a39a715a5c522686aad95026134be3d5731d06e/lib/utils.js#L124). For example, given a small trace (8% buffer usage), the raw inflated data <tt>Uint8Array</tt> contains 7,132,385 char codes which are _(1)_ converted into 109 strings of length 65,536 and subsequently _(2)_ joined into a single JSON string of length 7,132,385. By taking a DevTools heap dump right after the join (while the array of chunks is still in scope), we get the following results:


Object | Chars | Size | Size per char
-------- | ---------------- | ------- | -------------
concatenated JSON string | 7.1m | 25.9 MiB | 1.0 B
___single___ __JSON string chunk__ | 65.5k | 2.5 MiB | __40.0 B (!!!)__

The size of a single JSON string chunk is enormous. The reason for this is that V8 stores each chunk as a tree rather than a list (a chunk is selected in the screenshot below):

![screenshot from 2016-02-25 13 49 23](https://cloud.githubusercontent.com/assets/2546601/13321386/990c576a-dbc6-11e5-8566-43af40c2164e.png)

_(Note: As you can see in the screenshot above, chunks 0-100 are stored as flat strings and therefore have a size of 65600 (~1.0 B per characted). V8 seems to use the inefficient tree technique once there are more than 100 concatenated strings.)_

It turns out that <tt>String.fromCharCode(char0, char1, ..., char65535)</tt> is the same as <tt>String.fromCharCode(char0) + String.fromCharCode(char1) + ... + String.fromCharCode(char65535)</tt> [under the hood](https://code.google.com/p/chromium/codesearch#chromium/src/v8/src/js/string.js&l=567), so we cannot manually convert and concatenate individual characters either.

The next obvious candidate solution is to allocate a huge array (7,132,385 elements), store each converted character separately and then join the whole array (i.e. shrink chunk size to 1 byte). While this _does_ increase the size of traces that <tt>GzipImporter</tt> can handle (60% buffer usage imports without OOM), it still leads to OOM at full trace buffer usage. The reason for this is the large overhead of single-character strings in JS:

Object | Chars | Size | Size per char
-------- | ---------------- | ------- | -------------
concatenated JSON string | 7.1m | 25.9 MiB | 1.0 B
_single_ JSON string chunk | 65.5k | 2.5 MiB | 40.0 B (!!!)
___single character___ __string__ | 1 | 32.0 B | __32.0 B (!!!)__

Clearly, <tt>String.fromCharCode</tt> is not the way to go. Fortunately, Chrome provides a [<tt>TextDecoder</tt>](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) class which converts <tt>ArrayBuffer</tt> and <tt>ArrayBufferView</tt> objects to strings. Not only does this avoid OOM even at 100% buffer usage, it also [heavily outperforms the previous solutions in terms of speed](http://jsperf.com/arraybuffer-to-string-apply-performance/3):

Conversion approach | Ops per second
------------- | ---------------------
<tt>String.charCodeAt</tt> with 64 KiB chunks + <tt>Array.prototype.join</tt> | 18.59
<tt>String.charCodeAt</tt> with 1 B chunks + <tt>Array.prototype.join</tt> | 17.51
__<tt>TextDecoder.prototype.decode</tt>__ | __382 (yes, it's more than 20× faster!)__

I uploaded a small patch that replaces the <tt>JSZip.utils.transformTo</tt> call with a custom <tt>GzipImporter</tt> static method which uses <tt>TextDecoder</tt>: https://codereview.chromium.org/1739663003.

**Comment by natduca** [comment]
This is amaze. @paulirish fyi

**Comment by petrcermak** [comment]
Another case of the same OOM: https://bugs.chromium.org/p/chromium/issues/detail?id=587010

**Comment by petrcermak** [comment]
This should now be __fixed__ :white_check_mark:

**Comment by petrcermak** [comment]
I filed a v8 bug about this: https://bugs.chromium.org/p/v8/issues/detail?id=4786
